### PR TITLE
Don't run actions during evaluate

### LINKED ISSUE #720: Custom Actions are Executed during Rasa Core Evaluation
**Rasa Core version**: 0.9.7

**Python version**: 3.6

**Operating system** (windows, osx, ...): Docker (python:3.6-slim-jessie)

**Issue**:

When I run the Rasa Core evaluation, my custom action is run. I verified this by adding a `logging.error()` call to the action, which gets printed during evaluation.

In my case, this leads to misclassifications, because the action overwrites the slot values specified in the story file. But generally, custom actions could do anything, so they shouldn't be run during evaluation.

I run the evaluation as:
`python -m rasa_core.evaluate  --stories data/core --core models/current/dialogue --output evaluation/core_confusion_matrix.pdf --failed evaluation/failed_stories.txt`

Interestingly, `failed_stories.txt` says `All stories passed`, but the confusion matrix shows imperfect classifications.

### PR DESCRIPTION
**Pull request by tmbo**
**Proposed changes**:
- Don't run actions during evaluate
- fixed export format to include the whole stories 
- use the name of the original story if possible
- fixes #720 

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

### CODE REVIEW

**Comment by akelad** [comment]
Something i noticed, if there's an `OR` in a story, the story title becomes a repetition of the story title, separated by a `>` (e.g. `## just newsletter > just newsletter`) -- any way we can avoid this and just have the actual original title?

**Comment by akelad** [comment]
Also it prints out a failure for each OR possibility, one would be enough 😄

**Comment by tmbo** [comment]
I've pushed a fix for the naming, would be great if you can try it out. 

Concerning the number of failures: that is a tricky one and I don't think we can solve it as part of this one.

**Comment by akelad** [comment]
Haha yeah i thought that might be trickier. but yeah, will try out that fix!

**Comment by akelad** [comment]
just tested and the naming thing seems to be fixed :)

**Comment by tmbo** [comment]
ready for review ✅

File: rasa_core/evaluate.py [hunk 1]
```diff
@@ -58,149 +54,144 @@ def create_argument_parser():
...
+def collect_story_predictions(completed_trackers,
+                              agent,
+                              fail_on_prediction_errors=False):
+    """Test the stories from a file, running them through the stored model."""
 
-            preds.extend(last_prediction)
-            preds_padding = (len(actions_between_utterances) -
-                             len(last_prediction))
+    preds = []
+    gold = []
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: I'd name it `golds` as in above method

File: rasa_core/evaluate.py [hunk 2]
```diff
@@ -58,149 +54,144 @@ def create_argument_parser():
...
-        if last_prediction:
+def collect_story_predictions(completed_trackers,
+                              agent,
+                              fail_on_prediction_errors=False):
+    """Test the stories from a file, running them through the stored model."""
 
-            preds.extend(last_prediction)
-            preds_padding = (len(actions_between_utterances) -
-                             len(last_prediction))
+    preds = []
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: I'd name it `predictions` here or `preds` above

File: rasa_core/evaluate.py [hunk 3]
```diff
@@ -58,149 +54,144 @@ def create_argument_parser():
...
-            story["predicted"].extend(["None"] * preds_padding)
-            preds.extend(story["predicted"])
+    logger.info("Evaluating {} stories\n"
+                "Progress:".format(len(completed_trackers)))
 
-            actual.extend(actions_between_utterances)
-            actual_padding = (len(last_prediction) -
-                              len(actions_between_utterances))
+    for tracker in tqdm(completed_trackers):
+        curr_gold, curr_predictions, predicted_tracker = \
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: I don't like `curr`, can we rename it to `current`?

File: rasa_core/restore.py [hunk 4]
```diff
@@ -57,13 +58,49 @@ def _check_prediction_aligns_with_story(last_prediction,
...
 
-    p, a = evaluate.align_lists(last_prediction, actions_between_utterances)
+    p, a = align_lists(last_prediction, actions_between_utterances)
     if p != a:
         warnings.warn("Model predicted different actions than the "
                       "model used to create the story! Expected: "
                       "{} but got {}.".format(p, a))
 
 
+def align_lists(pred, actual):
```
Comment (Ghostvv) [inline rasa_core/restore.py]: let's name it `predictions`, `golds` here, or `preds`, `actuals`, but then rename `gold` to `actual` everywhere

File: tests/test_evaluation.py [hunk 5]
```diff
@@ -34,14 +33,11 @@ def test_evaluation_image_creation(tmpdir, default_agent):
 
 
 def test_evaluation_script(tmpdir, default_agent):
-    model_path = tmpdir.join("model").strpath
-    default_agent.persist(model_path)
+    completed_trackers = evaluate._generate_trackers(
+            DEFAULT_STORIES_FILE, default_agent)
 
     actual, preds, failed_stories = collect_story_predictions(
```
Comment (Ghostvv) [inline tests/test_evaluation.py]: consistent names here as wall

**Comment by tmbo** [comment]
Great push for better names :+1:

File: rasa_core/evaluate.py [hunk 6]
```diff
@@ -8,16 +8,13 @@
...
+from rasa_core import training, run
 from rasa_core import utils
 from rasa_core.agent import Agent
-from rasa_core.events import ActionExecuted, UserUttered
-from rasa_core.interpreter import RegexInterpreter, RasaNLUInterpreter
+from rasa_core.events import ActionExecuted
+from rasa_core.interpreter import NaturalLanguageInterpreter
 from rasa_core.trackers import DialogueStateTracker
 from rasa_core.training.generator import TrainingDataGenerator
 from rasa_nlu.evaluate import plot_confusion_matrix, log_evaluation_table
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: we probably should not import anything from `rasa_nlu`?

Comment (tmbo) [reply rasa_core/evaluate.py]: it's actually fine, rasa core depends on nlu (we also use things in other places from nlu) - the only things we shouldn't use are parts that depend on optional dependencies (e.g. spacy).

**Comment by Ghostvv** [comment]
I guess the last thing is: could you please add `number of correct stories out of` and then we are good to merge

File: rasa_core/evaluate.py [hunk 7]
```diff
@@ -140,25 +140,25 @@ def collect_story_predictions(completed_trackers,
...
+    predictions = []
+    golds = []
     failed = []
 
     logger.info("Evaluating {} stories\n"
                 "Progress:".format(len(completed_trackers)))
 
     for tracker in tqdm(completed_trackers):
-        curr_gold, curr_predictions, predicted_tracker = \
+        current_gold, current_predictions, predicted_tracker = \
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: should be `current_golds` I guess, because it is a list here

File: rasa_core/evaluate.py [hunk 8]
```diff
@@ -140,25 +140,25 @@ def collect_story_predictions(completed_trackers,
...
             _predict_tracker_actions(tracker, agent, fail_on_prediction_errors)
 
-        preds.extend(curr_predictions)
-        gold.extend(curr_gold)
+        predictions.extend(current_predictions)
+        golds.extend(current_gold)
 
-        if not curr_gold == curr_predictions:
+        if not current_gold == current_predictions:
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: `current_golds`

File: rasa_core/evaluate.py [hunk 9]
```diff
@@ -140,25 +140,25 @@ def collect_story_predictions(completed_trackers,
...
 
     for tracker in tqdm(completed_trackers):
-        curr_gold, curr_predictions, predicted_tracker = \
+        current_gold, current_predictions, predicted_tracker = \
             _predict_tracker_actions(tracker, agent, fail_on_prediction_errors)
 
-        preds.extend(curr_predictions)
-        gold.extend(curr_gold)
+        predictions.extend(current_predictions)
+        golds.extend(current_gold)
```
Comment (Ghostvv) [inline rasa_core/evaluate.py]: `current_golds`

**Comment by tmbo** [comment]
@Ghostvv all remarks have been addressed, this is ready to get merged if you give your ok

**Comment by Ghostvv** [comment]
sorry @tmbo, but I don't think you changed `current_gold` to `current_golds` 😅

File: rasa_core/evaluate.py (hunk 7 above)
Comment (Ghostvv) [reply rasa_core/evaluate.py]: current_gold to current_golds

File: rasa_core/evaluate.py (hunk 8 above)
Comment (Ghostvv) [reply rasa_core/evaluate.py]: current_gold to current_golds

File: rasa_core/evaluate.py (hunk 9 above)
Comment (Ghostvv) [reply rasa_core/evaluate.py]: golds.extend(current_golds)

**Comment by tmbo** [comment]
not sure what you mean, it is changed here: https://github.com/RasaHQ/rasa_core/pull/966/files#diff-4b73ffae216c7073d031c5bb15f53d11R157 is there any other location?

**Comment by Ghostvv** [comment]
yes, for some reason, it was displayed incorrectly

**Comment by tmbo** [comment]
so all good?

**Comment by tmbo** [comment]
That is printed now as well.

**Pull request merged**